// @infobip-api/sdk/src/models/2fa-models.ts
interface TwoFAApplication {
  configuration?: Configuration;
  enabled?: boolean;
  name: string;
}

interface Configuration {
  allowMultiplePinVerifications?: boolean;
  pinAttempts?: number;
  pinTimeToLive?: string;
  sendPinPerApplicationLimit?: string;
  sendPinPerPhoneNumberLimit?: string;
  verifyPinLimit?: string;
}

interface TwoFAMessageTemplate {
  language?: string;
  messageText: string;
  pinLength?: number;
  pinType: string;
  regional?: Regional;
  repeatDTMF?: string;
  senderId?: string;
  speechRate?: number;
}

interface Regional {
  indiaDlt?: IndiaDlt;
}

interface IndiaDlt {
  contentTemplateId?: string;
  principalEntityId: string;
}

interface TwoFAPinCode {
  applicationId: string;
  from?: string;
  messageId: string;
  placeholders?: object;
  to: string;
}

interface Pin {
  pin: string;
}

interface TwoFAVerificationStatus {
  msisdn: string;
  verified?: boolean;
  sent?: boolean;
}

export {
  TwoFAApplication,
  Configuration,
  TwoFAMessageTemplate,
  Regional,
  IndiaDlt,
  TwoFAPinCode,
  Pin,
  TwoFAVerificationStatus
}
